File: web/match_scoring.go
package web

import (
	"bytes"
	"net/http"
	"github.com/kennhung/ftcScoring/webTemplate"
	"github.com/kennhung/ftcScoring/model"
	"log"
	"github.com/kennhung/ftcScoring/play"
	"fmt"
	"io"
	"github.com/kennhung/ftcScoring/arena"
	"strings"
	"strconv"
)

func (web *Web) matchScoringHandler(w http.ResponseWriter, r *http.Request) {

	var data [3][]model.Match
	var err error

	currentMatch := web.arena.CurrentMatch
	data[0], err = web.arena.Database.GetMatchesByType("practice")
	if err != nil {
		handleWebErr(w, err)
		return
	}

	data[1], err = web.arena.Database.GetMatchesByType("qualification")
	if err != nil {
		handleWebErr(w, err)
		return
	}

	data[2], err = web.arena.Database.GetMatchesByType("elimination")
	if err != nil {
		handleWebErr(w, err)
		return
	}

	buffer := new(bytes.Buffer)
	template.Match_Scoring(data, currentMatch, buffer)

	w.Write(buffer.Bytes())
}

func (web *Web) matchScoringWebsocketHandler(w http.ResponseWriter, r *http.Request) {

	currentMatch := web.arena.CurrentMatch
	currentMatchResult, err := web.arena.Database.GetMatchResultForMatch(currentMatch.Id)
	if err != nil {
		log.Printf("Websocket error: %s", err)
		return
	} else if currentMatchResult == nil {
		//Not Saved yet
		currentMatchResult = model.NewMatchResult()
	}

	websocket, err := NewWebsocket(w, r)
	if err != nil {
		log.Printf("Websocket error: %s", err)
		return
	}

	defer websocket.Close()
	matchLoadTeamsListener := web.arena.MatchLoadTeamsChannel.Listen()
	defer close(matchLoadTeamsListener)
	matchTimeListener := web.arena.MatchTimeChannel.Listen()
	defer close(matchTimeListener)
	reloadDisplaysListener := web.arena.ReloadDisplaysChannel.Listen()
	defer close(reloadDisplaysListener)

	// Send the various notifications immediately upon connection.
	data := struct {
		RedScore  *play.Score
		BlueScore *play.Score
	}{web.arena.RedScore, web.arena.BlueScore}

	err = websocket.Write("score", data)
	if err != nil {
		log.Printf("Websocket error: %s", err)
		return
	}

	err = websocket.Write("matchTime", MatchTimeMessage{web.arena.MatchState, int(web.arena.MatchTimeSec())})
	if err != nil {
		log.Printf("Websocket error: %s", err)
		return
	}

	// Spin off a goroutine to listen for notifications and pass them on through the websocket.
	go func() {
		for {
			var messageType string
			var message interface{}
			select {
			case _, ok := <-matchLoadTeamsListener:
				if !ok {
					return
				}
				messageType = "reload"
				message = nil
			case matchTimeSec, ok := <-matchTimeListener:
				if !ok {
					return
				}
				messageType = "matchTime"
				message = MatchTimeMessage{web.arena.MatchState, matchTimeSec.(int)}
			case _, ok := <-reloadDisplaysListener:
				if !ok {
					return
				}
				messageType = "reload"
				message = nil
			}
			err = websocket.Write(messageType, message)
			if err != nil {
				// The client has probably closed the connection; nothing to do here.
				return
			}
		}
	}()

	for {
		messageType, data, err := websocket.Read()
		if err != nil {
			if err == io.EOF {
				// Client has closed the connection; nothing to do here.
				return
			}
			log.Printf("Websocket error: %s", err)
			return
		}

		switch messageType {
		case "commitMatch":
			if web.arena.MatchState != arena.PostMatch {
				// Don't allow committing the score until the match is over.
				websocket.WriteError("Cannot commit score: Match is not over.")
				continue
			}

			web.arena.ScoringStatusChannel.Notify(nil)
		case "updateRedScore":
			strs := strings.Split(fmt.Sprint(data),":")
			web.updateScore("Red",strs[0],strs[1])

		case "updateBlueScore":
			strs := strings.Split(fmt.Sprint(data),":")
			web.updateScore("Blue",strs[0],strs[1])

		default:
			websocket.WriteError(fmt.Sprintf("Invalid message type '%s'.", messageType))
			continue
		}

		// Send out the score again after handling the command, as it most likely changed as a result.
		data = struct {
			RedScore  *play.Score
			BlueScore *play.Score
		}{web.arena.RedScore,web.arena.BlueScore}
		err = websocket.Write("score", data)
		if err != nil {
			log.Printf("Websocket error: %s", err)
			return
		}
	}

}

func (web *Web) getCurrentMatchResult() *model.MatchResult {
	var RedCards map[string]string
	var BlueCards map[string]string

	if web.arena.Teams["R1"].YellowCard && web.arena.RedScore.Cards["R1"] == "Yellow" {
		RedCards["R1"] = "Red"
	} else if !web.arena.Teams["R1"].YellowCard && web.arena.RedScore.Cards["R1"] == "Yellow" {
		RedCards["R1"] = "Yellow"
	} else if web.arena.RedScore.Cards["R1"] == "Red" {
		RedCards["R1"] = "Red"
	}

	if web.arena.Teams["R2"].YellowCard && web.arena.RedScore.Cards["R2"] == "Yellow" {
		RedCards["R2"] = "Red"
	} else if !web.arena.Teams["R2"].YellowCard && web.arena.RedScore.Cards["R2"] == "Yellow" {
		RedCards["R2"] = "Yellow"
	} else if web.arena.RedScore.Cards["R2"] == "Red" {
		RedCards["R2"] = "Red"
	}
	if web.arena.Teams["B1"].YellowCard && web.arena.BlueScore.Cards["B1"] == "Yellow" {
		BlueCards["B1"] = "Red"
	} else if !web.arena.Teams["B1"].YellowCard && web.arena.BlueScore.Cards["B1"] == "Yellow" {
		BlueCards["B1"] = "Yellow"
	} else if web.arena.BlueScore.Cards["B1"] == "Red" {
		BlueCards["B1"] = "Red"
	}
	if web.arena.Teams["B2"].YellowCard && web.arena.BlueScore.Cards["B2"] == "Yellow" {
		BlueCards["B2"] = "Red"
	} else if !web.arena.Teams["B2"].YellowCard && web.arena.BlueScore.Cards["B2"] == "Yellow" {
		BlueCards["B2"] = "Yellow"
	} else if web.arena.BlueScore.Cards["B2"] == "Red" {
		BlueCards["B2"] = "Red"
	}

	return &model.MatchResult{MatchId: web.arena.CurrentMatch.Id, MatchType: web.arena.CurrentMatch.Type,
		RedScore: web.arena.RedScore, BlueScore: web.arena.BlueScore,
		RedCards: RedCards, BlueCards: BlueCards}
}

func (web *Web) updateScore(color string, field string, data string){
	score := new(play.Score)
	if color == "Red"{//Red
		score = web.arena.RedScore
	} else if color == "Blue"{//Blue
		score = web.arena.BlueScore
	}

	switch field {

	//Autonomous Period
	case "AutoJewels":
		score.AutoJewels,_ =  strconv.Atoi(data)
	case "AutoCryptobox":
		score.AutoCryptobox,_ = strconv.Atoi(data)
	case "CryptoboxKeys":
		score.CryptoboxKeys,_ = strconv.Atoi(data)
	case "RobotInSafeZone":
		score.RobotInSafeZone,_ = strconv.Atoi(data)

	//Driver-Controlled Period
	case "Glyphs":
		score.Glyphs,_ = strconv.Atoi(data)
	case "ComRows":
		score.ComRows,_ = strconv.Atoi(data)
	case "ComColumns":
		score.ComColumns,_ = strconv.Atoi(data)
	case "ComCiphers":
		score.ComCiphers,_ = strconv.Atoi(data)

	//End Game Period
	case "RelicsZ1":
		score.RelicsZ1,_ = strconv.Atoi(data)
	case "RelicsZ2":
		score.RelicsZ2,_ = strconv.Atoi(data)
	case "RelicsZ3":
		score.RelicsZ3,_ = strconv.Atoi(data)
	case "RelicsUpright":
		score.RelicsUpright,_ = strconv.Atoi(data)
	case "RobotBalanced":
		score.RobotBalanced,_ = strconv.Atoi(data)

	//Penalties
	case "MinorPena":
		score.Penalties[1],_ = strconv.Atoi(data)
	case "MajorPena":
		score.Penalties[0],_ = strconv.Atoi(data)
	}
}

File: res/js/page_scripts/match_scoring.js
var websocket;
var scoreCommitted = false;

function InitMaxandMin() {
    $(".numInput").attr("min", 0);
    $(".numInput").attr('value', 0);

    //Red
    //Autonomous Period
    $("#redScoreForm #AutoJewels").attr("max", 4);
    $("#redScoreForm #AutoCryptobox").attr("max", 48);
    $("#redScoreForm #CryptoboxKeys").attr("max", 2);
    $("#redScoreForm #RobotInSafeZone").attr("max", 2);

    //Driver-Controlled Period
    $("#redScoreForm #Glyphs").attr("max", 48)
    $("#redScoreForm #ComRows").attr("max", 8)
    $("#redScoreForm #ComColumns").attr("max", 6)
    $("#redScoreForm #ComCiphers").attr("max", 2)

    //End Game Period
    $("#redScoreForm #RelicsZ1").attr("max", 2)
    $("#redScoreForm #RelicsZ2").attr("max", 2)
    $("#redScoreForm #RelicsZ3").attr("max", 2)
    $("#redScoreForm #RelicsUpright").attr("max", 2)
    $("#redScoreForm #RobotBalanced").attr("max", 2)

    //Blue
    //Autonomous Period
    $("#blueScoreForm #AutoJewels").attr("max", 4);
    $("#blueScoreForm #AutoCryptobox").attr("max", 48);
    $("#blueScoreForm #CryptoboxKeys").attr("max", 2);
    $("#blueScoreForm #RobotInSafeZone").attr("max", 2);

    //Driver-Controlled Period
    $("#blueScoreForm #Glyphs").attr("max", 48)
    $("#blueScoreForm #ComRows").attr("max", 8)
    $("#blueScoreForm #ComColumns").attr("max", 6)
    $("#blueScoreForm #ComCiphers").attr("max", 2)

    //End Game Period
    $("#blueScoreForm #RelicsZ1").attr("max", 2)
    $("#blueScoreForm #RelicsZ2").attr("max", 2)
    $("#blueScoreForm #RelicsZ3").attr("max", 2)
    $("#blueScoreForm #RelicsUpright").attr("max", 2)
    $("#blueScoreForm #RobotBalanced").attr("max", 2)

}

var handleScore = function (data) {

    //Update RedScore
    var RedScore = data.RedScore;
    //Autonomous Period
    $("#redScoreForm #AutoJewels").val(RedScore.AutoJewels)
    $("#redScoreForm #AutoCryptobox").val(RedScore.AutoCryptobox)
    $("#redScoreForm #CryptoboxKeys").val(RedScore.CryptoboxKeys)
    $("#redScoreForm #RobotInSafeZone").val(RedScore.RobotInSafeZone)
    //Driver-Controlled Period
    $("#redScoreForm #Glyphs").val(RedScore.Glyphs)
    $("#redScoreForm #ComRows").val(RedScore.ComRows)
    $("#redScoreForm #ComColumns").val(RedScore.ComColumns)
    $("#redScoreForm #ComCiphers").val(RedScore.ComCiphers)
    //End Game Period
    $("#redScoreForm #RelicsZ1").val(RedScore.RelicsZ1)
    $("#redScoreForm #RelicsZ2").val(RedScore.RelicsZ2)
    $("#redScoreForm #RelicsZ3").val(RedScore.RelicsZ3)
    $("#redScoreForm #RelicsUpright").val(RedScore.RelicsUpright)
    $("#redScoreForm #RobotBalanced").val(RedScore.RobotBalanced)
    //Penalties
    $("#redScoreForm #MinorPena").val(RedScore.Penalties[1])
    $("#redScoreForm #MajorPena").val(RedScore.Penalties[0])

    //Update BlueScore
    var BlueScore = data.BlueScore;
    //Autonomous Period
    $("#blueScoreForm #AutoJewels").val(BlueScore.AutoJewels)
    $("#blueScoreForm #AutoCryptobox").val(BlueScore.AutoCryptobox)
    $("#blueScoreForm #CryptoboxKeys").val(BlueScore.CryptoboxKeys)
    $("#blueScoreForm #RobotInSafeZone").val(BlueScore.RobotInSafeZone)
    //Driver-Controlled Period
    $("#blueScoreForm #Glyphs").val(BlueScore.Glyphs)
    $("#blueScoreForm #ComRows").val(BlueScore.ComRows)
    $("#blueScoreForm #ComColumns").val(BlueScore.ComColumns)
    $("#blueScoreForm #ComCiphers").val(BlueScore.ComCiphers)
    //End Game Period
    $("#blueScoreForm #RelicsZ1").val(BlueScore.RelicsZ1)
    $("#blueScoreForm #RelicsZ2").val(BlueScore.RelicsZ2)
    $("#blueScoreForm #RelicsZ3").val(BlueScore.RelicsZ3)
    $("#blueScoreForm #RelicsUpright").val(BlueScore.RelicsUpright)
    $("#blueScoreForm #RobotBalanced").val(BlueScore.RobotBalanced)
    //Penalties
    $("#blueScoreForm #MinorPena").val(BlueScore.Penalties[1])
    $("#blueScoreForm #MajorPena").val(BlueScore.Penalties[0])

};

// Handles a websocket message to update the match status.
var handleMatchTime = function (data) {
    if (!scoreCommitted) {
        $("#scoringCard").show();
    } else {
        $("#scoringCard").hide();
    }
};


var commitMatchScore = function () {
    websocket.send("commitMatch");
};

$(function () {
    // Set up the websocket back to the server.
    websocket = new ScoringWebsocket("/match/scoring/websocket", {
        score: function (event) {
            handleScore(event.data);
        },
        matchTime: function (event) {
            handleMatchTime(event.data);
        }

    });

    $("#redScoreForm .numInput").change(function () {
        websocket.send("updateRedScore", $(this).attr("id") + ":" + $(this).val())
    });

    $("#blueScoreForm .numInput").change(function () {
        websocket.send("updateBlueScore", $(this).attr("id") + ":" + $(this).val())
    });

});
